impeller/renderer/backend/vulkan/binding_helpers_vk.cc

```diff
@@ -113,157 +122,111 @@ static bool BindBuffers(const Bindings& bindings,
     write_set.dstBinding = uniform.binding;
     write_set.descriptorCount = 1u;
     write_set.descriptorType = ToVKDescriptorType(layout.descriptor_type);
-    write_set.pBufferInfo = &buffers.back();
+    write_set.pBufferInfo = &buffer_workspace[buffer_offset - 1];
 
-    writes.push_back(write_set);
+    write_workspace[write_offset++] = write_set;
   }
   return true;
 }
 
-fml::StatusOr<std::vector<vk::DescriptorSet>> AllocateAndBindDescriptorSets(
+fml::StatusOr<vk::DescriptorSet> AllocateAndBindDescriptorSets(
     const ContextVK& context,
     const std::shared_ptr<CommandEncoderVK>& encoder,
-    const std::vector<Command>& commands,
-    const TextureVK& input_attachment) {
-  if (commands.empty()) {
-    return std::vector<vk::DescriptorSet>{};
-  }
-
-  // Step 1: Determine the total number of buffer and sampler descriptor
-  // sets required. Collect this information along with the layout information
-  // to allocate a correctly sized descriptor pool.
-  size_t buffer_count = 0;
-  size_t samplers_count = 0;
-  size_t subpass_count = 0;
-  std::vector<vk::DescriptorSetLayout> layouts;
-  layouts.reserve(commands.size());
-
-  for (const auto& command : commands) {
-    buffer_count += command.vertex_bindings.buffers.size();
-    buffer_count += command.fragment_bindings.buffers.size();
-    samplers_count += command.fragment_bindings.sampled_images.size();
-    subpass_count +=
-        command.pipeline->GetDescriptor().UsesSubpassInput() ? 1 : 0;
-
-    layouts.emplace_back(
-        PipelineVK::Cast(*command.pipeline).GetDescriptorSetLayout());
-  }
+    Allocator& allocator,
+    const Command& command,
+    const TextureVK& input_attachment,
+    std::array<vk::DescriptorImageInfo, kMaxBindings>& image_workspace,
+    std::array<vk::DescriptorBufferInfo, kMaxBindings>& buffer_workspace,
+    std::array<vk::WriteDescriptorSet, kMaxBindings + kMaxBindings>&
+        write_workspace) {
   auto descriptor_result = encoder->AllocateDescriptorSets(
-      buffer_count, samplers_count, subpass_count, layouts);
+      PipelineVK::Cast(*command.pipeline).GetDescriptorSetLayout(), context);
   if (!descriptor_result.ok()) {
     return descriptor_result.status();
   }
-  auto descriptor_sets = descriptor_result.value();
-  if (descriptor_sets.empty()) {
-    return fml::Status();
+  auto descriptor_set = descriptor_result.value();
+
+  size_t buffer_offset = 0u;
+  size_t image_offset = 0u;
+  size_t write_offset = 0u;
+
+  auto& pipeline_descriptor = command.pipeline->GetDescriptor();
+  auto& desc_set =
```

Comment:
missing the `&` on this meant we were always copying the descriptor set layouts read from the pipeline descriptor 🤦